File: modules/openlmis-web/src/main/java/org/openlmis/web/controller/ProductController.java
/*
 * This program is part of the OpenLMIS logistics management information system platform software.
 * Copyright © 2013 VillageReach
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *  
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.
 * You should have received a copy of the GNU Affero General Public License along with this program.  If not, see http://www.gnu.org/licenses.  For additional information contact info@OpenLMIS.org. 
 */

package org.openlmis.web.controller;

import org.openlmis.core.domain.*;
import org.openlmis.core.exception.DataException;
import org.openlmis.core.service.*;
import org.openlmis.core.web.controller.BaseController;
import org.openlmis.web.form.ProductDTO;
import org.openlmis.core.web.OpenLmisResponse;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;

import javax.servlet.http.HttpServletRequest;
import java.util.List;

import static org.openlmis.core.web.OpenLmisResponse.success;
import static org.springframework.http.HttpStatus.BAD_REQUEST;
import static org.springframework.web.bind.annotation.RequestMethod.*;

/**
 * This controller handles endpoint related to listing products.
 */
@RequestMapping(value = "/products")
@Controller
public class ProductController extends BaseController {

  @Autowired
  private ProductGroupService groupService;

  @Autowired
  private ProductFormService formService;

  @Autowired
  private ProductCategoryService productCategoryService;

  @Autowired
  private ProgramProductService programProductService;

  @Autowired
  private ProductPriceScheduleService priceScheduleService;

  @Autowired
  private ProductService service;

  @RequestMapping(value = "/groups", method = RequestMethod.GET, headers = ACCEPT_JSON)
  @PreAuthorize("@permissionEvaluator.hasPermission(principal,'MANAGE_PRODUCT')")
  public List<ProductGroup> getAllGroups() {
    return groupService.getAll();
  }

  @RequestMapping(value = "/forms", method = RequestMethod.GET, headers = ACCEPT_JSON)
  @PreAuthorize("@permissionEvaluator.hasPermission(principal,'MANAGE_PRODUCT')")
  public List<ProductForm> getAllForms() {
    return formService.getAll();
  }

  @RequestMapping(value = "/dosageUnits", method = RequestMethod.GET, headers = ACCEPT_JSON)
  @PreAuthorize("@permissionEvaluator.hasPermission(principal,'MANAGE_PRODUCT')")
  public List<DosageUnit> getAllDosageUnits() {
    return service.getAllDosageUnits();
  }

  @RequestMapping(value = "/{id}", method = GET, headers = ACCEPT_JSON)
  @PreAuthorize("@permissionEvaluator.hasPermission(principal,'MANAGE_PRODUCT')")
  public ProductDTO getById(@PathVariable(value = "id") Long id) {
    Product product = service.getById(id);

    if(product == null) return null;

    List<ProgramProduct> programProducts = programProductService.getByProductCode(product.getCode());

    List<ProductPriceSchedule> productPriceSchedules = priceScheduleService.getByProductId(product.getId());

    return new ProductDTO(product, product.getModifiedDate(), programProducts, productPriceSchedules);
  }

  @RequestMapping(method = POST, headers = ACCEPT_JSON)
  @PreAuthorize("@permissionEvaluator.hasPermission(principal,'MANAGE_PRODUCT')")
  public ResponseEntity<OpenLmisResponse> save(@RequestBody ProductDTO productDTO, HttpServletRequest request) {
    ResponseEntity<OpenLmisResponse> response;
    Product product = productDTO.getProduct();
    List<ProgramProduct> programProducts = productDTO.getProgramProducts();

    try {
      Long userId = loggedInUserId(request);
      product.setCreatedBy(userId);
      product.setModifiedBy(userId);
      service.save(product);
      programProductService.saveAll(programProducts, product);
    } catch (DataException e) {
      response = OpenLmisResponse.error(e, BAD_REQUEST);
      return response;
    }
    response = OpenLmisResponse.success(messageService.message("message.product.created.success", product.getName()));
    response.getBody().addData("productId", product.getId());
    return response;
  }

  @RequestMapping(value = "/{id}", method = PUT, headers = ACCEPT_JSON)
  @PreAuthorize("@permissionEvaluator.hasPermission(principal,'MANAGE_PRODUCT')")
  public ResponseEntity<OpenLmisResponse> update(@RequestBody ProductDTO productDTO, @PathVariable(value = "id") Long id,
                                                 HttpServletRequest request) {
    Product product = productDTO.getProduct();
    List<ProgramProduct> programProducts = productDTO.getProgramProducts();
    List<ProductPriceSchedule> productPriceSchedules = productDTO.getProductPriceSchedules();

    try {
      Long userId = loggedInUserId(request);
      product.setId(id);
      product.setModifiedBy(userId);
      service.save(product);
      programProductService.saveAll(programProducts, product);
      priceScheduleService.saveAll(productPriceSchedules, product);
    } catch (DataException e) {
      return OpenLmisResponse.error(e, BAD_REQUEST);
    }
    ResponseEntity<OpenLmisResponse> success = success(messageService.message("message.product.updated.success", product.getName()));
    success.getBody().addData("productId", product.getId());
    return success;
  }

  @RequestMapping(value = "/categories", method = RequestMethod.GET, headers = ACCEPT_JSON)
  @PreAuthorize("@permissionEvaluator.hasPermission(principal,'MANAGE_PRODUCT')")
  public List<ProductCategory> getAllCategories() {
    return productCategoryService.getAll();
  }
}

File: modules/openlmis-web/src/main/webapp/public/js/admin/product/controller/product-controller.js
/*
 * This program was produced for the U.S. Agency for International Development. It was prepared by the USAID | DELIVER PROJECT, Task Order 4. It is part of a project which utilizes code originally licensed under the terms of the Mozilla Public License (MPL) v2 and therefore is licensed under MPL v2 or later.
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the Mozilla Public License as published by the Mozilla Foundation, either version 2 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the Mozilla Public License for more details.
 *
 * You should have received a copy of the Mozilla Public License along with this program. If not, see http://www.mozilla.org/MPL/
 */

function ProductController($scope, $location, $dialog, messageService, ProductDetail , ReportPrograms, ProductList, RemoveProduct, RestoreProduct, DosageUnits, ProductForms) {

    $scope.title = 'Products';
    $scope.title = "Manage Products";

    // all products list
    ProductList.get({}, function (data) {
        $scope.filteredProducts = $scope.productsList = data.productList;
    }, function (data) {
        $location.path($scope.$parent.sourceUrl);
    });

    // apply filter
    $scope.filterProductsByProgram = function () {
        query = $scope.program;
        $scope.filteredProducts = [];
        query = query || "";
        if (query == 'NaN' || query === '') {
            $scope.filteredProducts = $scope.filterProductsByName($scope.productsList);
            return;
        }
        $scope.query = query;
        $scope.filteredArray = [];
        angular.forEach($scope.productsList, function (product) {
             $scope.product = product;
             angular.forEach($scope.product.programs, function (pp){
                 if(pp.id == $scope.query){
                     $scope.filteredArray.push($scope.product);
                 }
            });
        });
        $scope.filteredProducts =  $scope.filterProductsByName( $scope.filteredArray );
        $scope.resultCount = $scope.filteredProducts.length;
    };

    $scope.filterProductsByName = function(arrayOfProducts){
        if($scope.productName === '' || $scope.productName === undefined){
            return arrayOfProducts;
        }

        var result = [];
        angular.forEach(arrayOfProducts, function(product){
            if(product.primaryName.toLowerCase().indexOf($scope.productName.toLowerCase()) === 0){
                result.push(product);
            }
        });
        return result;

    };

    $scope.productLoaded = function () {
        return !($scope.products === undefined || $scope.products === null);
    };

    // Programs list
    ReportPrograms.get(function (data) {
        var tmp = data.programs;
        $scope.programs = data.programs;
    });

    $scope.YesNo = function (tf) {
        var retval = '';
        if (tf === true) {
            retval = 'Yes';
        } else
        {
            retval = 'No';
        }
        return retval;
    };

}